### PR TITLE
Add support for the osm-mysql interface

### PR DESCRIPTION
**Pull request by shayancanonical**
# Issue
The `osm` bundle uses the `osm-mariadb` charm, but would like to use this charm instead. However, the `mysql` relation interface (used by kubeflow) creates a restricted database user for use by the application. The `osm` bundle requires users with escalated privileges as it has an application (`osm-pol`) that creates a user and grants it privileges.

# Solution
Implement a separate `osm-mysql` relation interface, where the root and application user have escalated privileges to be able to create users and provide them with grants.

# Release Notes
Add support for the osm-mysql interface

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -68,11 +68,35 @@ commands =
...
+description = Run integration tests
+deps =
+    juju==2.9.11 # juju 3.3.0 has issues with retrieving action results
+    mysql-connector-python
+    pytest
+    pytest-operator
+    pytest-order
+    -r{toxinidir}/requirements.txt
+commands =
+    pytest -v --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs} --asyncio-mode=auto -m "ha_tests"
```
Comment (marceloneppel) [inline tox.ini:90]: Paulo had a [good suggestion](https://github.com/canonical/postgresql-operator/pull/10#discussion_r924885319) about moving the asyncio mode to `pyproject.toml` so it's possible to avoid specifying it here in each environment.

File: src/mysqlsh_helpers.py [hunk 2]
```diff
@@ -340,6 +356,129 @@ def remove_instances_not_online(self) -> None:
...
+
+    def delete_users_with_label(self, label_name: str, label_value: str) -> None:
+        """Delete users for a unit.
+
+        Args:
+            unit_name: The name of the unit for which to delete mysql users for
+
+        Raises:
+            MySQLDeleteUsersForUnitError if there is an error deleting users for the unit
+        """
```
Comment (marceloneppel) [inline src/mysqlsh_helpers.py:446]: The docstring needs to be updated.

**Review by marceloneppel — APPROVED** [review]
LGTM! Great job, Shayan!

File: src/relations/mysql.py [hunk 3]
```diff
@@ -142,6 +142,9 @@ def _on_mysql_relation_broken(self, event: RelationBrokenEvent) -> None:
         if not self.charm.unit.is_leader():
             return
 
+        if len(self.charm.model.relations[LEGACY_MYSQL]) > 1:
```
Comment (MiaAltieri) [inline src/relations/mysql.py:145]: add a comment for why you're doing this

File: src/relations/osm_mysql.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_created, self._on_osm_mysql_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_broken, self._on_osm_mysql_relation_broken
+        )
+
+    def _get_or_set_password_in_peer_databag(self, username: str) -> str:
```
Comment (MiaAltieri) [inline src/relations/osm_mysql.py:35]: `_get_password_in_peer_databag `

File: src/relations/osm_mysql.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        password = generate_random_password(PASSWORD_LENGTH)
+        peer_databag[f"{username}_password"] = password
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
```
Comment (MiaAltieri) [inline src/relations/osm_mysql.py:58]: Question: why store everything in the leaders databag rather than in the app data bag? Then you might not need t o copy things over all the time?

File: src/relations/osm_mysql.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
```
Comment (MiaAltieri) [inline src/relations/osm_mysql.py:95]: In order `self.charm._is_peer_data_set` the  ` on-config-changed ` event needs to run deferring here can cause this event to continuously defer making it take much longer for the config-changed hook to get on the queue before this event re-queues. This can lead to downtime/poor performance. Instead I suggest you set a flag and trigger and event that will re-call `_on_osm_mysql_relation_created`.

Mykola did something similar in his k8s charm [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/lib/charms/mongodb_libs/v0/mongodb_provider.py#L60-L63) and instead of re-triggering the event, he creating a function for himself to call once his condition was met by the start hook, see [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/src/charm.py#L142)

I think Paulo mentioned that this delay will only occur in a bundle but I think its worth looking out for.

File: src/relations/osm_mysql.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+        logger.warning("DEPRECATION WARNING - `osm-mysql` is a legacy interface")
+
+        username = self.charm.config.get("osm-mysql-interface-user")
+        database = self.charm.config.get("osm-mysql-interface-database")
+
+        if not username or not database:
```
Comment (MiaAltieri) [inline src/relations/osm_mysql.py:102]: comment for why

File: src/relations/osm_mysql.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    def _on_osm_mysql_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the 'mysql' legacy relation broken event.
+
+        Delete the application user created in the relation created
+        event handler.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if len(self.charm.model.relations[LEGACY_OSM_MYSQL]) > 1:
```
Comment (MiaAltieri) [inline src/relations/osm_mysql.py:148]: why?

File: tests/integration/test_charm.py [hunk 9]
```diff
@@ -127,6 +127,7 @@ async def test_consistent_data_replication_across_cluster(ops_test: OpsTest) ->
 
 @pytest.mark.order(3)
 @pytest.mark.abort_on_fail
+@pytest.mark.ha_tests
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:130]: might be nice to have. a different file since there will be _a lot_ of HA tests, but I am not opinionated on this right now :)

File: tests/integration/test_osm_mysql.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                        assert "keystone" in output
+                        assert "pol" in output
+
+                        # test that osm-pol successfully creates tables
+                        output = await execute_queries_on_unit(
+                            unit_address,
+                            server_config_credentials["username"],
+                            server_config_credentials["password"],
+                            get_count_pol_tables,
+                        )
```
Comment (MiaAltieri) [inline tests/integration/test_osm_mysql.py:158]: might be worth checking delete and update functions

File: tests/integration/test_osm_mysql.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+                        assert output[0] > 0
+        except RetryError:
+            assert False
```
Comment (MiaAltieri) [inline tests/integration/test_osm_mysql.py:161]: might be worth checking that it has "users with escalated privileges as it has an application (osm-pol) that creates a user and grants it privileges." since this is what makes this relation unique.

File: tests/integration/test_osm_mysql.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        await ops_test.model.block_until(lambda: is_relation_joined(ops_test, "mongodb", "mongo"))
+
+        await ops_test.model.relate("osm-pol:kafka", "osm-kafka:kafka")
+        await ops_test.model.block_until(lambda: is_relation_joined(ops_test, "kafka", "kafka"))
+
+        await ops_test.model.relate("osm-pol:mysql", f"{APP_NAME}:osm-mysql")
+        await ops_test.model.block_until(
+            lambda: is_relation_joined(ops_test, "mysql", "osm-mysql")
+        )
+
```
Comment (MiaAltieri) [inline tests/integration/test_osm_mysql.py:123]: I think it would be nice to break this into two tests

`test_deploy_and_relate` and `test_osm_pol_ops`

File: tox.ini [hunk 13]
```diff
@@ -68,11 +68,35 @@ commands =
...
+
+[testenv:integration-osm-mysql]
```
Comment (MiaAltieri) [inline tox.ini:92]: be sure you add this test to `.github/workflows/release.yaml` otherwise it won't run these tests before releasing to charmhub

**Review by MiaAltieri — COMMENTED** [review]
Looking AWESOME Shayan! 🤩

File: src/relations/osm_mysql.py (hunk 5 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py:58]: The issue is that reactive charms (which the osm charms are) read relation data from the leader unit's databag. The `leader-elected` event handler exists to make sure that if a new leader is elected, the relation data is copied from the app databag to the new leader's databag

File: src/relations/osm_mysql.py (hunk 6 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py:95]: I'm not sure that this idea will work in this case. There are 2 conditions that must be met before the event handler (function) can be called:
1. The `on-config-changed` should have completed execution at least once
2. The `osm-mysql` relation must have been created
With these conditions and machinery at hand, I believe that the existing approach is sufficient (and I agree with Paulo that this will only happen once in a bundle at the time of relation being created)

File: tests/integration/test_osm_mysql.py (hunk 11 above)
Comment (shayancanonical) [reply tests/integration/test_osm_mysql.py:161]: After double-checking the application (osm-pol), the existence of the `pol` database and tables within it implies that the escalated user has been created correctly (as osm-pol requires privileges to create a user to be able to run the migrations). Additionally, it's hard to test for the existence of a user created by osm-pol as its naming scheme is not straight-forward

File: tests/integration/test_osm_mysql.py (hunk 10 above)
Comment (shayancanonical) [reply tests/integration/test_osm_mysql.py:158]: I'm unsure which user we should test update + delete grants for. Ideally, it would be the user created by the osm-pol, but we are unable to guess the username that the application creates. Instead, here we just check that the application can run migrations (and creates tables)

File: tox.ini (hunk 1 above)
Comment (shayancanonical) [reply tox.ini]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/mysqlsh_helpers.py (hunk 2 above)
Comment (shayancanonical) [reply src/mysqlsh_helpers.py]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/osm_mysql.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            return
+
+        logger.warning("DEPRECATION WARNING - `osm-mysql` is a legacy interface")
+
+        self.charm._mysql.delete_users_with_label("label", "osm-mysql-legacy-relation")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/osm_mysql.py]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/osm_mysql.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            event.relation.data[self.charm.unit].update(updates)
+
+            return
+
+        password = self._get_or_set_password_in_peer_databag(username)
+
+        self.charm._mysql.create_database(database)
+        self.charm._mysql.create_user(username, password, "osm-mysql-legacy-relation")
+        self.charm._mysql.escalate_user_privileges("root")
+        self.charm._mysql.escalate_user_privileges(username)
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/osm_mysql.py]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/mysql.py (hunk 3 above)
Comment (shayancanonical) [reply src/relations/mysql.py:176]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/osm_mysql.py (hunk 4 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/osm_mysql.py (hunk 7 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py:119]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/relations/osm_mysql.py (hunk 8 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py:183]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: tests/integration/test_charm.py (hunk 9 above)
Comment (shayancanonical) [reply tests/integration/test_charm.py:130]: we will address this comment in a different PR when the need arises (when we have added more HA tests)

File: tests/integration/test_osm_mysql.py (hunk 12 above)
Comment (shayancanonical) [reply tests/integration/test_osm_mysql.py:123]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: tox.ini (hunk 13 above)
Comment (shayancanonical) [reply tox.ini:92]: addressed in 2f6e6e434f2bab7f2a508153eb234fb73edf2bed

File: src/mysqlsh_helpers.py (hunk 2 above)
Comment (marceloneppel) [reply src/mysqlsh_helpers.py]: It seems that it was not updated in that commit.

Comment (shayancanonical) [reply src/mysqlsh_helpers.py]: i believe i updated it [here](https://github.com/canonical/mysql-k8s-operator/pull/88/files#diff-6e566a7b93efc3f3e7ac0e6d8989256725f876f06af78c45d0ae3815d4bcf82eR454) -> the name of the function is slightly different than the copy pasted one

Comment (marceloneppel) [reply src/mysqlsh_helpers.py]: Ah, sorry Shayan! I haven't made it clear that more parts of the docstring need changes (the args section and maybe the initial description can be updated; the [args section](https://github.com/canonical/mysql-k8s-operator/commit/2f6e6e434f2bab7f2a508153eb234fb73edf2bed#diff-6e566a7b93efc3f3e7ac0e6d8989256725f876f06af78c45d0ae3815d4bcf82eR451) is not matching the list of arguments of this method)

Comment (shayancanonical) [reply src/mysqlsh_helpers.py]: ah sorry, i completely missed that. corrected in f375b063af5362d0b622dce487e68f2bfd54a630

**Review by marceloneppel — APPROVED** [review]

File: src/relations/mysql.py (hunk 3 above)
Comment (paulomach) [reply src/relations/mysql.py:175]: `s/osm-mysql/mysql/` ?

File: src/relations/osm_mysql.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+        user_exists = False
+        try:
+            user_exists = self.charm._mysql.does_mysql_user_exist(username, "%")
+        except MySQLCheckUserExistenceError:
+            self.charm.unit.status = BlockedStatus("Failed to check user existence")
+            return
+
+        # Only execute if the application user does not exist
+        if user_exists:
```
Comment (paulomach) [inline src/relations/osm_mysql.py:130]: If i understood correctly, this case will be valid when the joining unit is from a second application being related using `osm-mysql`.
right? or am I out to lunch?

File: tests/integration/test_osm_mysql.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        )
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications["osm-zookeeper"].units) == 1
+        )
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications["osm-mongodb"].units) == 1
+        )
+
+        await ops_test.model.relate("osm-kafka:zookeeper", "osm-zookeeper:zookeeper")
+        await ops_test.model.block_until(
```
Comment (paulomach) [inline tests/integration/test_osm_mysql.py:91]: Consider adding `timeout`'s to all `block_until` calls.

File: tox.ini [hunk 18]
```diff
@@ -68,11 +68,35 @@ commands =
 [testenv:integration]
 description = Run integration tests
 deps =
-    juju
+    juju==2.9.11 # juju 3.3.0 has issues with retrieving action results
```
Comment (paulomach) [inline tox.ini:71]: did you meant `2.9.33` ? (the current stable in 2.x series)

**Review by paulomach — APPROVED** [review]
Looking smooth - some minor comments to take a peak!

File: src/relations/osm_mysql.py (hunk 16 above)
Comment (shayancanonical) [reply src/relations/osm_mysql.py:130]: you're correct. we cannot prevent multiple applications from being related on the same interface/endpoint. this check is just address this edge case

File: tests/integration/test_osm_mysql.py (hunk 17 above)
Comment (shayancanonical) [reply tests/integration/test_osm_mysql.py:96]: added in [750358c](https://github.com/canonical/mysql-k8s-operator/pull/88/commits/750358c2a892641c312e45f948db18bc9efe8924)

File: tox.ini (hunk 18 above)
Comment (shayancanonical) [reply tox.ini:71]: the highest 2.9 version for python-libjuju is `2.9.11` (i tried to change it to `2.9.33`, but got an error)
